### PR TITLE
User-supplied reencryption conditions

### PR DESCRIPTION
**Pull request by KPrasch**
**Type of PR:**
Feature

**Required reviews:** 
1

**What this does:**
- Introduces framework for the expression and evaluation of arbitrary on-chain reencryption preconditions expressed as JSON.
- Decryption conditions are defined by `Enrico` and attached to a `MessageKit` to implement per-encryption-level conditions.
- Ursula evaluates conditions defined for ethereum (via `RPCCondition`, `ContractCondition`)
- nucypher core shims for `ReencryptionRequest`, `MessageKit`, `RetrievalKit`
- Modifies Bob's retrieval flow to handle conditions (as `ConditionLingo`) associated with work order capsules.

**Notes for reviewers:**
This PR is targeting an epic.  The remaining test failures are caused because of backwards-incompatible changes that will be farther advanced by nucypher/nucypher-core #2989.

### CODE REVIEW

**Comment by jMyles** [comment]
:eyes:

File: nucypher/policy/conditions/evm.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+            real_name = re.sub(r'(?<!^)(?=[A-Z])', '_', real_name).lower()
+            p = kwargs.get(real_name)
+            if not p:
+                raise Exception(f'"{real_name}" not found in request context')
+        processed_parameters.append(p)
+    return processed_parameters
+
+
+class RPCCondition(ReencryptionCondition):
+    ALLOWED_METHODS = (  # TODO: Deny list instead of allow list, if any?
```
Comment (theref) [inline nucypher/policy/conditions/evm.py:89]: The only method I think we'd want to Deny is `[eth_sendRawTransaction](https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_sendrawtransaction)`

File: tests/acceptance/blockchain/conditions/test_conditions.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    assert result is True
+
+
+def test_time_condition_evaluation(testerchain, timelock_condition):
+    address = testerchain.unassigned_accounts[0]
+    result, value = timelock_condition.verify(provider=testerchain.provider, user_address=address)
+    assert result is True
+
+
+def test_simple_compound_conditions_evaluation(testerchain):
```
Comment (theref) [inline tests/acceptance/blockchain/conditions/test_conditions.py:35]: are we able to handle nested conidtions? (condition 1 AND condition 2) OR condition 3?

**Comment by derekpierre** [comment]
@KPrasch - this is amazing so far.

File: tests/acceptance/blockchain/conditions/test_conditions.py (hunk 2 above)
Comment (KPrasch) [reply tests/acceptance/blockchain/conditions/test_conditions.py:35]: Presently, no.

File: nucypher/policy/conditions/time.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        name = fields.Str()
+        method = fields.Str(default='timelock')
+        return_value_test = fields.Nested(ReturnValueTest.ReturnValueTestSchema())
+
+        @post_load
+        def make(self, data, **kwargs):
+            return TimeCondition(**data)
+
+    def __repr__(self) -> str:
+        r = f'{self.__class__.__name__}(timestamp={self.return_value_test.value})'
```
Comment (derekpierre) [inline nucypher/policy/conditions/time.py:23]: Should the '=' in 'timestamp=' really be the comparator on the return_value_test? i.e.
```suggestion
        r = f'{self.__class__.__name__}(timestamp{self.return_value_test.comparator}{self.return_value_test.value})'
```

File: nucypher/core.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        payload = bytes(self._core_instance)
+        if self.condition:
+            payload += self._DELIMITER
+            payload += bytes(self.condition)
+        return payload
+
+    @classmethod
+    def from_bytes(cls, data: bytes):
+        condition = None
+        if cls._DELIMITER in data:
```
[the remark this answers is not in the document]
Comment (fjarri) [reply nucypher/core.py:50]: `.partition()` would probably be more idiomatic

**Review by derekpierre — APPROVED** [review]
🎸

**Pull request merged**